runtime/android/core_internal/src/org/xwalk/core/internal/XWalkWebResourceResponseInternal.java

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) 2015 Intel Corporation. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package org.xwalk.core.internal;
+
+import android.util.Log;
+import org.chromium.base.CalledByNative;
+import org.chromium.base.JNINamespace;
+
+import java.io.InputStream;
+import java.io.StringBufferInputStream;
+import java.util.Map;
+
+/**
+ * Encapsulates a resource response. Applications can return an instance of this
+ * class from {@link XWalkResourceClient#shouldInterceptRequest} to provide a custom
+ * response when the XWalkView requests a particular resource.
+ */
+@JNINamespace("xwalk")
+@XWalkAPI(createInternally = true)
+public class XWalkWebResourceResponseInternal{
+    private String mMimeType;
+    private String mEncoding;
+    private InputStream mData;
+    private int mStatusCode;
+    private String mReasonPhrase;
+    private Map<String, String> mResponseHeaders;
+    private String[] mResponseHeaderNames;
+    private String[] mResponseHeaderValues;
+
+    // Never use this constructor.
+    // It is only used in XWalkWebResourceResponseBridge.
+    XWalkWebResourceResponseInternal() {
+        mMimeType = null;
+        mEncoding = null;
+        mData = null;
+    }
+
+    XWalkWebResourceResponseInternal(String mimeType, String encoding, InputStream data) {
+        mMimeType = mimeType;
+        mEncoding = encoding;
+        setData(data);
+    }
+
+    XWalkWebResourceResponseInternal(String mimeType, String encoding, InputStream data,
+            int statusCode, String reasonPhrase, Map<String, String> responseHeaders) {
+        this(mimeType, encoding, data);
+
+        mStatusCode = statusCode;
+        mReasonPhrase = reasonPhrase;
+        mResponseHeaders = responseHeaders;
+    }
+
+    private void fillInResponseHeaderNamesAndValuesIfNeeded() {
+        if (mResponseHeaders == null || mResponseHeaderNames != null) return;
+        mResponseHeaderNames = new String[mResponseHeaders.size()];
+        mResponseHeaderValues = new String[mResponseHeaders.size()];
+        int i = 0;
+        for (Map.Entry<String, String> entry : mResponseHeaders.entrySet()) {
+            mResponseHeaderNames[i] = entry.getKey();
+            mResponseHeaderValues[i] = entry.getValue();
+            i++;
+        }
+    }
+
+    /**
+     * Sets the resource response's MIME type, for example &quot;text/html&quot;.
+     *
+     * @param mimeType The resource response's MIME type
+     * @since 6.0
+     */
+    @XWalkAPI
+    public void setMimeType(String mimeType) {
+        mMimeType = mimeType;
+    }
+
+    /**
+     * Gets the resource response's MIME type.
+     *
+     * @return The resource response's MIME type
+     * @since 6.0
+     */
+    @XWalkAPI
+    public String getMimeType() {
+        return mMimeType;
+    }
+
+    @CalledByNative
+    public String getMimeTypeNative() {
+        return mMimeType;
+    }
```

Comment:
it would be better to implement with ``` return getMimeType()``` as you just want to provide a different function name for ```@CalledByNative``` related handling.